src/python/zquantum/qcbm/cost_function.py

```diff
@@ -84,13 +49,25 @@ def _evaluate(self, parameters: np.ndarray) -> (float, BitstringDistribution):
             (float): cost function value for given parameters
             zquantum.core.bitstring_distribution.BitstringDistribution: distribution obtained
         """
-        circuit = self.ansatz.get_executable_circuit(parameters)
-        distribution = self.backend.get_bitstring_distribution(circuit)
+        circuit = ansatz.get_executable_circuit(parameters)
+        distribution = backend.get_bitstring_distribution(circuit)
         value = evaluate_distribution_distance(
-            self.target_bitstring_distribution,
+            target_bitstring_distribution,
             distribution,
-            self.distance_measure,
-            distance_measure_parameters=self.distance_measure_parameters,
+            distance_measure,
+            distance_measure_parameters=distance_measure_parameters,
+        )
+
+        if store_artifact:
+            store_artifact("bitstring_distribution", distribution)
+
+        return ValueEstimate(value)
+
+    if gradient_type == "finite_difference":
+        cost_function = function_with_gradient(
+            cost_function, finite_differences_gradient(cost_function)
         )
+    else:
+        raise RuntimeError("Unsupported gradient type: ", gradient_type)
```

Comment:
We haven't discussed it yet, but I have some doubts regarding where the initialization of the gradient should be placed. Let's leave it for now though because it probably needs some careful planning.

Reply:
Fair point - I agree that this isn't ideal, but better to leave it for now and change when we're sure.